opal/datatype/opal_convertor.c

```diff
@@ -324,6 +324,205 @@ int32_t opal_convertor_unpack( opal_convertor_t* pConv,
     return pConv->fAdvance( pConv, iov, out_size, max_data );
 }
 
+int32_t
+opal_iovec_set_position( opal_convertor_t *convertor, size_t *position )
+{
+    if( 0 == *position ) {
+        convertor->pStack[0].count = convertor->count;
+        convertor->pStack[0].disp = 0;
+        convertor->pStack[1].index = 0;
+        convertor->pStack[1].disp = 0;
+        convertor->bConverted = 0;
+
+        return 0;
+    }
+
+    const opal_datatype_t *pData = convertor->pDesc;
+    size_t done_count = *position / pData->size;
+
+    struct iovec *iov = pData->iov;
+    uint32_t i = 0;
+
+    convertor->bConverted = *position;
+    convertor->pStack[0].count = convertor->count - done_count;
+
+    size_t track = *position % pData->size;
+    for( i = 0; i < pData->iovcnt; i++ ){
+        if( track < iov[i].iov_len )
+            break;
+        track -= iov[i].iov_len;
+    }
+
+    convertor->pStack[1].disp = track;
+    convertor->pStack[1].index = i;
+    convertor->pStack[0].disp = done_count * (pData->ub - pData->lb);
+
+    return 0;
+}
+
+int32_t
+opal_iovec_pack( opal_convertor_t *convertor,
+        struct iovec *out_iov,
+        uint32_t *out_size,
+        size_t *max_data )
+{
+    const opal_datatype_t *pData = convertor->pDesc;
+    struct iovec *iov = pData->iov;
+    char *dst = out_iov[0].iov_base,
+         *src = convertor->pBaseBuf + convertor->pStack[0].disp;
+    size_t track = *max_data, iov_track;
+    uint32_t i, iov_count = 0;
+
+    dst = out_iov[iov_count].iov_base;
+    iov_track = out_iov[iov_count].iov_len;
+
+    while( convertor->pStack[0].count ){
+        for( i = convertor->pStack[1].index; i < pData->iovcnt; i++ ) {
+            if( track == 0 ){
+                convertor->pStack[1].index = i;
+                goto complete_pack;
+            }
+
+restart_pack:
+            if( iov_track < (iov[i].iov_len - convertor->pStack[1].disp) && iov_track < track ){
```

Comment:
I still think that this if and the one below at line 405 can and should be merged.